### PR TITLE
Update `tests/accelerators/*.py` to use `devices` instead of `gpus` or `ipus`

### PR DESCRIPTION
**Pull request by mathemusician**
## What does this PR do?

Update test to deprecate `num_processes`, `gpus`, `tpu_cores`, and `ipus` from the Trainer constructor

Part of [Pull 11040](https://github.com/PyTorchLightning/pytorch-lightning/pull/11040).

🌳 [Trello Board Link](https://trello.com/b/He49BfCc/pl-tests)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by daniellepintz** [comment]
@mathemusician what is the status of this PR?

**Comment by mathemusician** [comment]
Still working on it

**Comment by four4fish** [comment]
Hey @mathemusician  we have a 1.6 release and code freeze start from next week, this is an blocker for https://github.com/PyTorchLightning/pytorch-lightning/pull/11040. Anything I can help with? It will be great if we can have this merged early next week

**Comment by mathemusician** [comment]
Hi @four4fish ! I think any help with the tpu code would be awesome!

**Comment by mathemusician** [comment]
Finished with rebase. Will check back in tomorrow to work on this again.

**Comment by daniellepintz** [comment]
@mathemusician thanks for your persistence on this. I have an idea that may make things easier. For all the tests that are failing right now, can you just remove the changes to those tests so they succeed, and we will handle those in a different PR? 

We want to keep some of the old tests anyway, to make sure the deprecated functionality is still working.

**Comment by mathemusician** [comment]
> @mathemusician thanks for your persistence on this. I have an idea that may make things easier. For all the tests that are failing right now, can you just remove the changes to those tests so they succeed, and we will handle those in a different PR?
> 
> We want to keep some of the old tests anyway, to make sure the deprecated functionality is still working.

@daniellepintz That may be best recourse. I sometimes bite off more than I can chew.

**Comment by daniellepintz** [comment]
No worries, it's a huge task! Thanks.

**Comment by daniellepintz** [comment]
FYI the GPUs failure is due to https://github.com/PyTorchLightning/pytorch-lightning/issues/12314

File: tests/accelerators/test_accelerator_connector.py [hunk 1]
```diff
@@ -546,27 +527,17 @@ def test_accelerator_gpu_with_devices(devices, plugin):
...
-    assert trainer.gpus == devices
+    assert trainer.devices == devices
     assert isinstance(trainer.strategy, plugin)
     assert isinstance(trainer.accelerator, GPUAccelerator)
 
 
 @RunIf(min_gpus=1)
 def test_accelerator_auto_with_devices_gpu():
     trainer = Trainer(accelerator="auto", devices=1)
-    assert isinstance(trainer.accelerator, GPUAccelerator)
```
Comment (akihironitta) [inline tests/accelerators/test_accelerator_connector.py:557]: We should keep the accelerator instance type check, no?

**Review by akihironitta — COMMENTED** [review]
@mathemusician Thank you for your work! Left a few minor comments. Could you have a look?

File: tests/accelerators/test_accelerator_connector.py [hunk 2]
```diff
@@ -99,6 +100,7 @@ def test_accelerator_choice_ddp_spawn(cuda_available_mock, device_count_mock):
 )
 @mock.patch("torch.cuda.set_device")
 @mock.patch("torch.cuda.device_count", return_value=2)
+@mock.patch("torch.cuda.is_available", return_value=True)
```
[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:103]: Yup, deleting

File: tests/accelerators/test_accelerator_connector.py [hunk 3]
```diff
@@ -125,6 +127,7 @@ def test_accelerator_choice_ddp_slurm(*_):
...
```
[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:130]: Yup, deleting

File: tests/accelerators/test_accelerator_connector.py [hunk 4]
```diff
@@ -470,12 +473,12 @@ def test_plugin_accelerator_choice(accelerator: Optional[str], plugin: str):
...
 @mock.patch("torch.cuda.device_count", return_value=2)
-@pytest.mark.parametrize("gpus", [1, 2])
+@pytest.mark.parametrize("devices", [1, 2])
 def test_accelerator_choice_multi_node_gpu(
-    mock_is_available, mock_device_count, tmpdir, accelerator: str, plugin: ParallelStrategy, gpus: int
+    mock_is_available, mock_device_count, tmpdir, accelerator: str, plugin: ParallelStrategy, devices: int
 ):
     with pytest.deprecated_call(match=r"accelerator=.*\)` has been deprecated"):
-        trainer = Trainer(accelerator=accelerator, default_root_dir=tmpdir, num_nodes=2, gpus=gpus)
+        trainer = Trainer(default_root_dir=tmpdir, num_nodes=2, accelerator=accelerator, gpus=devices)
```
[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:481]: wow, good catch

File: tests/accelerators/test_accelerator_connector.py (hunk 1 above)
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:557]: Yup

Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:557]: I'll revert the test.

**Review by four4fish — APPROVED** [review]

File: tests/accelerators/test_accelerator_connector.py [hunk 5]
```diff
@@ -454,7 +455,7 @@ def test_plugin_accelerator_choice(accelerator: Optional[str], plugin: str):
     assert isinstance(trainer.strategy, DDPShardedStrategy)
 
     with pytest.deprecated_call(match="Passing .* `strategy` to the `plugins`"):
-        trainer = Trainer(plugins=plugin, num_processes=2)
+        trainer = Trainer(plugins=plugin, accelerator="cpu", num_processes=2)
```
Comment (awaelchli) [inline tests/accelerators/test_accelerator_connector.py:458]: what about this?

File: tests/accelerators/test_accelerator_connector.py [hunk 6]
```diff
@@ -814,10 +777,12 @@ def test_strategy_choice_ddp_te(*_):
 )
 @mock.patch("torch.cuda.set_device")
 @mock.patch("torch.cuda.device_count", return_value=2)
+@mock.patch("torch.cuda.is_available", return_value=True)
+@mock.patch("torch.cuda.is_available", return_value=True)
```
Comment (awaelchli) [inline tests/accelerators/test_accelerator_connector.py:781]: duplicated

**Review by awaelchli — APPROVED** [review]

File: tests/accelerators/test_accelerator_connector.py (hunk 5 above)
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:458]: Wow, yeah, I’m gonna change that

File: tests/accelerators/test_accelerator_connector.py (hunk 6 above)
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:781]: Good catch

File: tests/accelerators/test_accelerator_connector.py (hunk 5 above)
Comment (mathemusician) [reply tests/accelerators/test_accelerator_connector.py:458]: Wait, that might be one of the reverted tests. *testing*

Comment (awaelchli) [reply tests/accelerators/test_accelerator_connector.py]: yeah, that's what I meant :))

**Comment by daniellepintz** [comment]
@mathemusician Mind rebasing? Also looks like `test_accelerator_gpu_with_devices` is failing

File: tests/accelerators/test_accelerator_connector.py [hunk 7]
```diff
@@ -556,7 +535,7 @@ def test_accelerator_gpu_with_devices(devices, plugin):
 
     trainer = Trainer(accelerator="gpu", devices=devices)
 
-    assert trainer.gpus == devices
+    assert trainer.num_devices == devices
```
Comment (rohitgr7) [inline tests/accelerators/test_accelerator_connector.py:538]:
```suggestion
    assert trainer.num_devices == len(devices) if isinstance(devices, list) else devices
```

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
